crates/scmi/src/devices/common.rs

```diff
@@ -0,0 +1,550 @@
+    );
+    devices.insert(
+        String::from("iio"),
+        DeviceSpecification::new(
+            iio::IIOSensor::new_device,
+            String::from("industrial I/O sensor"),
+            String::from(""),
+            vec![
+                String::from(
+                    "path: path to the device directory (e.g. /sys/bus/iio/devices/iio:device0)",
+                ),
+                String::from("channel: prefix of the device type (e.g. in_accel)"),
+                String::from("name: an optional name of the sensor, max. 15 characters"),
+            ],
+        ),
+    );
+    devices
+}
+
+fn devices_help() -> String {
+    let mut help = String::new();
+    writeln!(help, "Available devices:").unwrap();
+    for (name, specification) in available_devices().iter() {
+        let short_help = &specification.short_help;
+        let long_help = &specification.long_help;
+        let parameters_help = &specification.parameters_help;
+        writeln!(help, "\n- {name}: {short_help}").unwrap();
+        for line in long_help.lines() {
+            writeln!(help, "  {line}").unwrap();
+        }
+        if !parameters_help.is_empty() {
+            writeln!(help, "  Parameters:").unwrap();
+            for parameter in parameters_help {
+                writeln!(help, "  - {parameter}").unwrap();
+            }
+        }
+    }
+    writeln!(help, "\nDevice specification example:").unwrap();
+    writeln!(
+        help,
+        "--device iio,path=/sys/bus/iio/devices/iio:device0,channel=in_accel"
+    )
+    .unwrap();
+    help
+}
+
+pub(crate) fn print_devices_help() {
+    let help = devices_help();
+    println!("{}", help);
+}
+
+pub(crate) fn check_device_properties(
+    properties: &DeviceProperties,
+    required: &[String],
+    optional: &[String],
+) -> Result<(), String> {
+    let missing = properties.missing(required);
+    if !missing.is_empty() {
+        return Err(format!(
+            "Missing device parameters: {}",
+            missing.iter().sorted().join(", ")
+        ));
+    }
+    let mut all_allowed = Vec::from(required);
+    all_allowed.extend(optional.iter().map(String::from));
+    let extra = properties.extra(&all_allowed);
+    if !extra.is_empty() {
+        return Err(format!(
+            "Unexpected device parameters: {}",
+            extra.iter().sorted().join(", ")
+        ));
+    }
+    Ok(())
+}
+
+// Common sensor infrastructure
+
+/// Basic information about the sensor.
+///
+/// It is typically used as a field in structs implementing sensor devices.
+pub(crate) struct Sensor {
+    /// The sensor name (possibly truncated) as reported to the guest.
+    pub(crate) name: String,
+    /// Whether the sensor is enabled.
+    ///
+    /// Sensors can be enabled and disabled using SCMI.  [Sensor]s created
+    /// using [Sensor::new] are disabled initially.
+    enabled: bool,
+}
+
+impl Sensor {
+    pub(crate) fn new(properties: &DeviceProperties, default_name: &String) -> Sensor {
```

Comment:
```suggestion
    pub(crate) fn new(properties: &DeviceProperties, default_name: &str) -> Sensor {
```
as you are calling to_string at the end anyway